apps/zotonic_core/src/models/m_identity.erl

```diff
@@ -1191,62 +1191,60 @@ insert(Rsc, Type, Key, Context) ->
 insert(Rsc, Type, Key, Props, Context) ->
     KeyNorm = normalize_key(Type, Key),
     case is_valid_key(Type, KeyNorm, Context) of
-        true -> insert_1(Rsc, Type, KeyNorm, Props, Context);
-        false -> {error, invalid_key}
+        true ->
+            F = fun(Ctx) ->
+                        insert_1(Rsc, Type, KeyNorm, Props, Ctx)
+                end,
+            z_db:transaction(F, Context);
```

Comment:
Shouldn't we check here for an error like:  `#error{ codename = unique_violation }` ?
And if so, check if it is because of the (Id, Type, Key) constraint?

Reply:
I tried that. When you catch the `unique_violoation` and retry the transaction it will also fail, because there is another concurrent transaction which is touching the same row.

Reply:
Ah, the locking problem where we can't lock the key for update yet, because it doesn't exist yet.

Either a table lock, backing off for a couple of msec, or a transactional lock is needed in these situations.
I will check the options.  In the paysub module I am now using a retry and a "select ... for update", which seems to work fine for now, but I see the problem with the race condition, especially if the insert takes longer.

Reply:
Forgot, since PostgreSQL 9.5 the "INSERT ... ON CONFLICT .." exists.  Maybe this is a good moment to start using it in the core:

https://www.postgresql.org/docs/9.5/sql-insert.html

As the oldest supported PostgreSQL version is now 11 we might want to up our minimum PostgreSQL version to 9.5?

Reply:
Ah... we could extend `z_db:insert/3` with some options...

Something like this?

```erlang
z_db:insert(identity, #{ rsc_id => ..., ...}, 
            #{ on_conflict => <<"ON CONSTAINT(identity_rsc_id_type_key_unique)">>,
               action => <<"DO NOTHING">> },
            Context),
```

Reply:
In this case we want to merge the existing props.

Reply:
Merging would only work correctly when the only properties are real column, and not something which ends up in a props column.

Reply:
I was thinking in two different directions.  One is an option to handle the conflict with an update   `{on_conflict, update}`.  Second option is to add an extra function `upsert`, next to `insert` and `update`.

Reply:
Sounds logical... At the moment this would be impossible for erlang props columns of course. Maybe something for 1.1? We could make something to merge json props then.

Reply:
Yes, and in the meantime we can just have a "last one wins" strategy.